### PR TITLE
fix: `builtin:swc-loader` does not report error correctly

### PR DESCRIPTION
**Pull request by h-a-n-a**
## Summary

Continue https://github.com/web-infra-dev/rspack/pull/4082

In webpack, `emitWarning` or `emitError` can be used to create some additional error on the compilation. 

In Rspack, we have the `loader_context.diagnostics` field to represent this behavior. But this one has been misused in builtin-loaders. In order to make error report successfully and treat the loader error as the module build error, I changed it to returning a `Err`. 

But still, there's some misalignment in the `ModuleBuildError`, I will create a new issue to indicate this problem.

## Test Plan

Added.

### CODE REVIEW

**Review by ahabhgk — DISMISSED** [review]

File: package.json [hunk 1]
```diff
@@ -24,7 +24,7 @@
     "prepare": "is-ci || husky install",
     "pnpm:devPreinstall": "is-ci || node ./scripts/doctor.js",
     "test:example": "pnpm --filter \"example-*\" build",
-    "test:unit": "pnpm --filter \"@rspack/*\" test",
+    "test:unit": "NO_COLOR=1 pnpm --filter \"@rspack/*\" test",
```
Comment (ahabhgk) [inline package.json:27]: You can use `stats.colors = false` to omit the color

Comment (ahabhgk) [reply package.json:27]: Can not control swc's internal error emitter

Comment (h-a-n-a) [reply package.json:27]: Ok, sounds that is going to make this PR not focused, it's move on to the next one for that.

Comment (h-a-n-a) [reply package.json:27]: Seems https://github.com/web-infra-dev/rspack/issues/4124 they don't have this issue, have to figure out why.

**Review by ahabhgk — APPROVED** [review]

File: package.json (hunk 1 above)
Comment (hardfist) [reply package.json:27]: https://github.com/swc-project/swc/blob/4623b3e07f9cedb9652785f977f2bd0e509981d9/crates/swc_error_reporters/src/handler.rs#L53 swc's internal error can be controlled here

**Pull request merged**